pkg/steps/ip_pool.go

```diff
@@ -105,9 +110,70 @@ func (s *ipPoolStep) run(ctx context.Context) error {
 		s.ipPoolLease.resources = names
 	}
 
+	if len(names) > 0 {
+		go s.checkAndReleaseUnusedLeases(ctx, minute)
+	}
+
 	wrappedErr := results.ForReason("executing_test").ForError(s.wrapped.Run(ctx))
 	logrus.Infof("Releasing ip pool leases for test %s", s.Name())
 	releaseErr := results.ForReason("releasing_ip_pool_lease").ForError(releaseLeases(client, *l))
 
 	return aggregateWrappedErrorAndReleaseError(wrappedErr, releaseErr)
 }
+
+const UnusedIpCount = "UNUSED_IP_COUNT"
+
+// checkAndReleaseUnusedLeases periodically checks for a positive value in the
+// UNUSED_IP_COUNT data in the shared secret which signals that there are leases that can be released.
+// If/when this is discovered it will release that number of leases, and stop checking.
+// minute is provided as an argument to assist with unit testing.
+func (s *ipPoolStep) checkAndReleaseUnusedLeases(ctx context.Context, minute time.Duration) {
```

Comment:
Can we add extra debug log to understand the average wait time for an ipPool Lease? This will help us calibrate it better and avoid longer wait times.

Reply:
We don't wait at all, if they aren't available immediately we move on. The 15 minute timeout is just for checking if the `UNUSED_IP_COUNT` file is present, this will be set by the workflow step, and will likely be present in the first minute or two. This just provides a buffer in case the install step takes longer than anticipated.

I am still working through the fix to the race here, but the idea with this function is that it is async and lightweight. It won't affect test times at all.